### PR TITLE
[#766] Move data from sendgrid to firestore

### PR DESCRIPTION
**Pull request by jhflorey**
This PR address issue https://github.com/episphere/connect/issues/766

- Added the keyName of GCP Secret-Manager that is storing the verification keys of the sendgrid event webhook  into env files.
- Added endpoint `sendGridEventWebhook` for the sendgrid event webhook
- Installed package `@sendgrid/eventwebhook` to verify  the sendgrid event webhook verification key.
- Added `custom_args` to send email parameters to send some data (connect_id, notification_id, token) to sendgrid. We will get it back while receiving data from event webhook
- Added function `processEventWebhook` in firestore.js for processing data that received from event webhook

### CODE REVIEW

File: utils/sendGridEventWebhook.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                // {
+                //     to: "",
+                //     custom_args: {
+                //         connect_id: "cnid2",
+                //         token: "tk2",
+                //         notification_id: "nid2",
+                //         gcloud_project: process.env.GCLOUD_PROJECT,
+                //         attempt: "1",
+                //     },
+                // },
```
Comment (Gbarra9) [inline utils/sendGridEventWebhook.js:36]: Is this commented out code needed?

File: utils/sendGridEventWebhook.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        await sgMail.send(msg);
+        console.log("Complete sending email");
+        return res.status(200).json({ code: 200 });
+    } catch (e) {
+        console.error(e);
+        return res.status(500).json({ code: 500 });
+    }
+};
+
+const _receivedEvents = async (req, res) => {
```
Comment (Gbarra9) [inline utils/sendGridEventWebhook.js:64]: For my understanding, what is the meaning of an underscore before the variable name?

**Review by Gbarra9 — APPROVED** [review]
Looks good.

File: utils/sendGridEventWebhook.js (hunk 1 above)
Comment (jhflorey) [reply utils/sendGridEventWebhook.js:36]: It's used for testing send to multiple emails

File: utils/sendGridEventWebhook.js (hunk 2 above)
Comment (jhflorey) [reply utils/sendGridEventWebhook.js:64]: I would like to mark it as an internal function used in the file (not export)

File: utils/sendGridEventWebhook.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+const { EventWebhook, EventWebhookHeader } = require("@sendgrid/eventwebhook");
+const { processEventWebhook } = require("./firestore");
+const { getResponseJSON } = require("./shared");
+
+const _testSendEmail = async (res) => {
+    console.log("Start sending email");
+
+    try {
+        // Add SendGrid apiKey to test
+        sgMail.setApiKey("");
```
[the remark this answers is not in the document]
Comment (jhflorey) [reply utils/sendGridEventWebhook.js:12]: We must have the permission to get it in SendGrid dashboard

File: utils/firestore.js [hunk 4]
```diff
@@ -2809,6 +2809,52 @@ const getSpecimensByCollectionIds = async (collectionIdsArray, siteCode, isBPTL
...
+            notification_id: event.notification_id,
+            sg_event_id: event.sg_event_id,
+            sg_message_id: event.sg_message_id,
+            token: event.token,
+        };
+        if (["bounce", "dropped"].includes(event.event)) {
+            eventRecord[`${event.event}_reason`] = event.reason;
+        }
+        await db.collection("sendgridTracking").add(eventRecord);
+    }
```
Comment (JoeArmani) [inline utils/firestore.js:2855]: Looks like we can try/catch here and throw -> _receivedEvents

File: utils/sendGridEventWebhook.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        // Update data to test
+        const msg = {
+            personalizations: [
+                {
+                    to: "",
```
Comment (JoeArmani) [inline utils/sendGridEventWebhook.js:18]: Is this field used in production?

File: utils/sendGridEventWebhook.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        );
+
+        if (verification) {
+            await _receivedEvents(req, res);
+        } else {
+            res.status(403).send("Forbidden");
+        }
+    } else return res.status(400).json(getResponseJSON("Bad request!", 400));
+};
+
```
Comment (JoeArmani) [inline utils/sendGridEventWebhook.js:117]: probably good to wrap this with try/catch, too, and return 500 on error.

**Review by JoeArmani — APPROVED** [review]
looks good, just a couple comments and a question.

File: utils/sendGridEventWebhook.js (hunk 6 above)
Comment (jhflorey) [reply utils/sendGridEventWebhook.js:117]: it makes sense.

Comment (jhflorey) [reply utils/sendGridEventWebhook.js]: Solved

**Review by JoeArmani — APPROVED** [review]
looks good

**Pull request merged**